src/irr/asset/CImageLoaderTGA.cpp

```diff
@@ -112,78 +112,129 @@ bool CImageLoaderTGA::isALoadableFileFormat(io::IReadFile* _file) const
+		region.imageExtent = imgInfo.extent;
+
+		inputCreationImage = asset::ICPUImage::create(std::move(imgInfo));
+		inputCreationImage->setBufferAndRegions(std::move(texelBuffer), regions);
+	}
 
-	auto channels = getFormatChannelCount(destFormat);
-	auto stride = regionBufferRowLenght * getTexelOrBlockBytesize(destFormat);
-	
-	if (flip)
-		out += size.X * size.Y * channels;
-	
-	for (int y = 0; y < size.Y; ++y) {
-		if (flip)
+	core::smart_refctd_ptr<ICPUImage> newConvertedImage;
+	{
+		auto referenceImageParams = inputCreationImage->getCreationParameters();
+		auto referenceBuffer = inputCreationImage->getBuffer();
+		auto referenceRegions = inputCreationImage->getRegions();
+		auto referenceRegion = referenceRegions.begin();
+		const auto newTexelOrBlockByteSize = asset::getTexelOrBlockBytesize(destFormat);
+
+		asset::TexelBlockInfo referenceBlockInfo(referenceImageParams.format);
+		core::vector3du32_SIMD referenceTrueExtent = referenceBlockInfo.convertTexelsToBlocks(referenceRegion->getTexelStrides());
+
+		auto newImageParams = referenceImageParams;
+		auto newCpuBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(referenceTrueExtent.X * referenceTrueExtent.Y * referenceTrueExtent.Z * newTexelOrBlockByteSize);
+		auto newRegions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1);
+		newRegions->front() = *referenceRegion;
+
+		newImageParams.format = destFormat;
+		newConvertedImage = ICPUImage::create(std::move(newImageParams));
+		newConvertedImage->setBufferAndRegions(std::move(newCpuBuffer), newRegions);
+	}
+
+	auto attachedRegion = newConvertedImage->getRegions().begin();
+
+	state.inImage = inputCreationImage.get();
+	state.outImage = newConvertedImage.get();
+	state.inOffset = { 0, 0, 0 };
+	state.inBaseLayer = 0;
+	state.outOffset = { 0, 0, 0 };
+	state.outBaseLayer = 0;
+	state.extent = attachedRegion->getExtent();
+	state.layerCount = attachedRegion->imageSubresource.layerCount;
+	state.inMipLevel = attachedRegion->imageSubresource.mipLevel;
+	state.outMipLevel = attachedRegion->imageSubresource.mipLevel;
+
+	if (!convertFilter.execute(&state))
+		os::Printer::log("Something went wrong while converting!", ELL_WARNING);
+
+	bool OpenGlFlip = !flip;
+	if (OpenGlFlip)
+	{
+		auto newFormat = newConvertedImage->getCreationParameters().format;
+		asset::TexelBlockInfo blockInfo(newFormat);
+		core::vector3du32_SIMD trueExtent = blockInfo.convertTexelsToBlocks(newConvertedImage->getRegions().begin()->getTexelStrides());
+
+		auto buffer = newConvertedImage->getBuffer();
+		auto out = reinterpret_cast<uint8_t*>(buffer->getPointer()) + buffer->getSize();
+		auto stride = trueExtent.X * getTexelOrBlockBytesize(newFormat);
+
+		for (uint32_t y = 0; y < trueExtent.Y; ++y)
 			out -= stride;
-		
-		const void *src_container[4] = {in, nullptr, nullptr, nullptr};
-		convertColor<srcFormat, destFormat>(src_container, out, stride, size);
-		in += stride;
-		
-		if (!flip)
-			out += stride;
 	}
```

Comment:
there's no flipping going on here, just pointer arithmetic

Reply:
it does what it did before I think and seems it flips image

Reply:
cause rendered is flipped in reference to written one, but actaully you are right probably

Reply:
fixed and tested

Reply:
it's done